### PR TITLE
iOS support

### PR DESCRIPTION
**Pull request by jeremie-movify**
Apart of the added iOS code, I updated the backgound channel's method name from "backgroundChannelInitialized" to "backgroundJobDidComplete" ; that's open for discussion.
I'll also update the README for detailed iOS instructions in a later commit.

### CODE REVIEW

File: android/src/main/kotlin/be/tramckrijte/workmanager/EchoingWorker.kt [hunk 1]
```diff
@@ -29,7 +29,7 @@ class EchoingWorker(private val ctx: Context,
         const val IS_IN_DEBUG_MODE = "be.tramckrijte.workmanager.IS_IN_DEBUG_MODE"
 
         const val BACKGROUND_CHANNEL_NAME = "be.tramckrijte.workmanager/background_channel_work_manager"
-        const val BACKGROUND_CHANNEL_INITIALIZED = "backgroundChannelInitialized"
+        const val BACKGROUND_CHANNEL_INITIALIZED = "backgroundJobDidComplete"
```
Comment (timrijckaert) [inline android/src/main/kotlin/be/tramckrijte/workmanager/EchoingWorker.kt]: Even thought the name is completely internal so it does not really matter what this is called, I think it implies the implementer code was already done executing while it has merely started the background channel. 
We would still need to send the execution result back on this channel.

File: example/lib/main.dart [hunk 2]
```diff
@@ -11,8 +13,10 @@ const simplePeriodicTask = "simplePeriodicTask";
 const simplePeriodic1HourTask = "simplePeriodic1HourTask";
 
 void callbackDispatcher() {
+  stderr.writeln('Debug - presumably performing a network request');
+
   Workmanager.defaultCallbackDispatcher((echoValue) {
-    print("Native echoed: $echoValue");
```
Comment (timrijckaert) [inline example/lib/main.dart]: Revert this change

File: example/lib/main.dart [hunk 3]
```diff
@@ -11,8 +13,10 @@ const simplePeriodicTask = "simplePeriodicTask";
 const simplePeriodic1HourTask = "simplePeriodic1HourTask";
 
 void callbackDispatcher() {
+  stderr.writeln('Debug - presumably performing a network request');
```
Comment (timrijckaert) [inline example/lib/main.dart]: Revert this change

File: example/lib/main.dart [hunk 4]
```diff
@@ -111,6 +115,13 @@ class _MyAppState extends State<MyApp> {
                       frequency: Duration(hours: 1),
                     );
                   }),
+              Text("iOS", style: Theme.of(context).textTheme.headline),
```
Comment (timrijckaert) [inline example/lib/main.dart]: I think I would definitely emphasise that this is for debugging purposes only.
Even debatable if you really need to expose it to the user. 

If we do opt to have this debugging function. I would check that this only works when `_isInDebugMode` is enabled.

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 5]
```diff
@@ -1,14 +1,135 @@
...
+        registrar.addMethodCallDelegate(instance, channel: foregroundMethodChannel)
+        registrar.addApplicationDelegate(instance)
+    }
+    
+    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
+        
+        switch call.method {
+        case ForegroundMethodChannel.methods.initialize.rawValue:
+            guard let callbackHandle: Int64 = call.arguments as? Int64 else {
+                result(WMPError.unexpectedMethodArguments(argumentsDescription: call.arguments.debugDescription))
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: This does not seem necessary. 
It should always emit an int.

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 6]
```diff
@@ -1,14 +1,135 @@
...
+                return
+            }
+            store(callbackHandle)
+            result(nil)
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: Maybe do `result(true)` then it is the same as on Android

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 7]
```diff
@@ -1,14 +1,135 @@
...
+        case ForegroundMethodChannel.methods.performSampleFetch.rawValue:
+            triggerSampleBackgroundFetch()
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: Debatable as explained above.

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 8]
```diff
@@ -1,14 +1,135 @@
...
+
+//MARK: - AppDelegate conformance
+
+extension SwiftWorkmanagerPlugin {
+    
+    override public func application(_ application: UIApplication, performFetchWithCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void) -> Bool {
+        
+        // First, let's retreive our callBack handle
+        guard let callbackHandle: Int64 = getStoredCallbackHandle() else {
+            completionHandler(.failed)
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: I would go for the most optimistic approach here. Just forcefully unwrap it. If it crashes then fine. It should.
Now we are just swallowing that error.

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 9]
```diff
@@ -1,14 +1,135 @@
...
+        let flutterCallbackInformation: FlutterCallbackInformation = FlutterCallbackCache.lookupCallbackInformation(callbackHandle)
+        let flutterEngine = FlutterEngine.init(name: flutterThreadLabelPrefix, project: nil, allowHeadlessExecution: true)!
+        flutterEngine.run(withEntrypoint: flutterCallbackInformation.callbackName, libraryURI: flutterCallbackInformation.callbackLibraryPath)
+        
+        // Since we're now running a specific Flutter engine, no MethodChannel exists ; let's create one for WorkManager's BackgroundMethodChannel
+        let backgroundMethodChannel = FlutterMethodChannel(name: BackgroundMethodChannel.channelName, binaryMessenger: flutterEngine)
+        backgroundMethodChannel.setMethodCallHandler { (call, result) in
+            guard
+                call.method == BackgroundMethodChannel.methods.backgroundJobDidComplete.rawValue,
+                let argument: UInt = call.arguments as? UInt,
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: Why is there an int sent here?

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 10]
```diff
@@ -1,14 +1,135 @@
...
+                else {
+                    result(WMPError.unexpectedMethodArguments(argumentsDescription: call.arguments.debugDescription))
+                    return
+            }
+            // Method name and arguments match the expected values ; call the native completionHandler
+            result(nil)
+            completionHandler(backgroundFetchResult)
+        }
+        
+        return true
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: What does this imply?
`true` as in, the background fetch was executed successfully?  

We need to wait for the answer from Dart side and return its value here.

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 11]
```diff
@@ -1,14 +1,135 @@
...
+    var callbackHandleStorageKey: String {
+        return "callBackHandleStorageKey"
+    }
+    
+    func store(_ callbackHandle: Int64) {
+        UserDefaults.standard.set(callbackHandle, forKey: callbackHandleStorageKey)
+    }
+    
+    func getStoredCallbackHandle() -> Int64? {
+        return UserDefaults.standard.value(forKey: callbackHandleStorageKey) as? Int64
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: Forcefully unwrap it

File: lib/workmanager.dart [hunk 12]
```diff
@@ -112,7 +115,8 @@ class Workmanager {
   }) async {
     Workmanager._isInDebugMode = isInDebugMode;
     final callback = PluginUtilities.getCallbackHandle(callbackDispatcher);
-    await _foregroundChannel.invokeMethod('initialize', callback.toRawHandle());
+    final callbackHandle = callback.toRawHandle();
```
Comment (timrijckaert) [inline lib/workmanager.dart]: Revert

File: lib/workmanager.dart [hunk 13]
```diff
@@ -98,9 +99,11 @@ class Workmanager {
   static void defaultCallbackDispatcher(
       final EchoCallbackFunction echoFunction) {
     WidgetsFlutterBinding.ensureInitialized();
-    _backgroundChannel
-        .setMethodCallHandler((call) async => echoFunction(call.arguments));
-    _backgroundChannel.invokeMethod("backgroundChannelInitialized");
+    _backgroundChannel.setMethodCallHandler((call) async {
+      stderr.writeln('DEBUG');
```
Comment (timrijckaert) [inline lib/workmanager.dart]: Revert

File: lib/workmanager.dart [hunk 14]
```diff
@@ -215,4 +219,9 @@ class Workmanager {
   /// Cancels all jobs
   static Future<void> cancelAll() async =>
       await _foregroundChannel.invokeMethod("cancelAll");
+
+  static Future<void> onIOSPerformFetch() async {
+    print("Debug - WorkManager - onIOSPerformFetch");
```
Comment (timrijckaert) [inline lib/workmanager.dart]: Debatable like explained above.  
Wrap with `_isInDebug`

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 15]
```diff
@@ -1,14 +1,135 @@
...
-    let instance = SwiftWorkmanagerPlugin()
-    registrar.addMethodCallDelegate(instance, channel: channel)
-  }
+public class SwiftWorkmanagerPlugin: FlutterPluginAppLifeCycleDelegate {
+    
+    private struct ForegroundMethodChannel {
+        static let channelName = "be.tramckrijte.workmanager/foreground_channel_work_manager"
+        enum methods: String {
+            case initialize
+            case performSampleFetch
```
[the remark this answers is not in the document]
Comment (timrijckaert) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: Or we could use the OptionalMethodChannel in Dart.

```
/// A [MethodChannel] that ignores missing platform plugins.
///
/// When [invokeMethod] fails to find the platform plugin, it returns null
/// instead of throwing an exception.
```

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 16]
```diff
@@ -1,14 +1,135 @@
...
+                call.method == BackgroundMethodChannel.methods.backgroundJobDidComplete.rawValue,
+                let argument: UInt = call.arguments as? UInt,
+                let backgroundFetchResult = UIBackgroundFetchResult.init(rawValue: argument)
+                else {
+                    result(WMPError.unexpectedMethodArguments(argumentsDescription: call.arguments.debugDescription))
+                    return
+            }
+            // Method name and arguments match the expected values ; call the native completionHandler
+            result(nil)
+            completionHandler(backgroundFetchResult)
```
[the remark this answers is not in the document]
Comment (petermnt) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: Correct, after receiving the `backgroundChannelInitialized` event (which was incorrectly renamed to backgroundJobDidComplete) it should call the `performIosFetch` method on the channel, and use the result of that future to call the completionHandler

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 17]
```diff
@@ -1,14 +1,137 @@
...
+                        let backgroundFetchResult = UIBackgroundFetchResult.init(rawValue: UInt(fetchResult))
+                        else {
+                            result(WMPError.unexpectedMethodArguments(flutterResult.debugDescription))
+                            completionHandler(.failed)
+                            return
+                    }
+                    
+                    result(true)
+                    completionHandler(backgroundFetchResult)
+                })
```
Comment (petermnt) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: result(true) should be here

File: README.md [hunk 18]
```diff
@@ -47,17 +36,18 @@ void callbackDispatcher() {
 }
 
 Workmanager.initialize(
-    callbackDispatcher, //the top level function.
-    isInDebugMode: true //If enabled it will post a notificiation whenever the job is running. Handy for debugging jobs
+    callbackDispatcher, // The top level function, aka Flutter entry point
```
Comment (timrijckaert) [inline README.md]: > aka Flutter entry point

Not really.

File: README.md [hunk 19]
```diff
@@ -74,12 +64,11 @@ Workmanager.registerPeriodicTask(
 )
 ```
 
-You will need to provide a unique name; this comes in handy when you want to cancel this task later on.  
-The second parameter is the `String` that will be returned to your `callbackDispatcher` function.  
-You can use this `String` to identify which work needs to be done.  
+Each task must have a **unique name** ; this allows cancellation of a started task.  
+The second parameter is the `String` that will be sent to your `callbackDispatcher` function, indicating the task's *type*.  
 
 ## Customisation
```
Comment (timrijckaert) [inline README.md]: Maybe we should have the Customisation header under the Android usage header, so it's more clear this can only work for Android

File: README.md [hunk 20]
```diff
@@ -126,32 +115,58 @@ Workmanager.registerOneOffTask(
...
-### All
+- ##### cancel all registered tasks
 
 ```
 Workmanager.cancelAll();
 ```
+
+## iOS usage
+
+Background taks on iOS are very different. Before anything, make sure you've added the following key to your project's `Info.plist :
```
Comment (timrijckaert) [inline README.md]: Is this something that you typically do in XCode or just by hand?
Otherwise maybe a screenshot? I always place these in a `.art` folder in the root of the project

File: README.md [hunk 21]
```diff
@@ -126,32 +115,58 @@ Workmanager.registerOneOffTask(
...
+    <string>fetch</string>
+  </array>
+</key>
+```
+
+#### Set the application's minimumBackgroundFetchInterval
+
+Set your desired *minimumBackgroundFetchInterval* in your app's delegate's `didFinishLaunchingWithOptions` :
+
+`UIApplication.shared.setMinimumBackgroundFetchInterval(TimeInterval(60 * 15)) // E.g. max every 15 minutes`
```
Comment (timrijckaert) [inline README.md]: max? minimum right?

File: README.md [hunk 22]
```diff
@@ -126,32 +115,58 @@ Workmanager.registerOneOffTask(
...
+> Note : this time interval is a minimum ; there's no guarantee about how ofteh this will be called. 
+
+#### Adding the plugin as an AppDelegate
+
+This is done automatically by the SwiftWorkManagerPlugin at registration time, by calling [addApplicationDelegate](https://api.flutter.dev/objcdoc/Protocols/FlutterPluginRegistrar.html#/c:objc(pl)FlutterPluginRegistrar(im)addApplicationDelegate:) on the Flutter registrar.
```
Comment (timrijckaert) [inline README.md]: Is it necessary to inform the user of this automatic action? 
I think it can only be confusing, no?

File: README.md [hunk 23]
```diff
@@ -126,32 +115,58 @@ Workmanager.registerOneOffTask(
...
+
+#### Waiting for iOS to trigger `performFetchWithCompletionHandler`
+
+We don't have any control on how often iOS will allow our app to fetch data in the background. The Example Flutter project allows to simulate this event, by pressing the 'Simulate Background Fetch' button (since Xcode's Debug > Simulate Background Fetch doesn't).
```
Comment (timrijckaert) [inline README.md]: 👍

File: example/lib/main.dart [hunk 24]
```diff
@@ -13,23 +14,19 @@ const simplePeriodicTask = "simplePeriodicTask";
...
 void callbackDispatcher() {
-  stderr.writeln('Debug - presumably performing a network request');
-
   Workmanager.defaultCallbackDispatcher((echoValue) {
-    stderr.writeln('Debug - Native echoed: $echoValue');
-
     switch (echoValue) {
       case simpleTaskKey:
-        print("$simpleTaskKey was executed");
+        stderr.writeln("$simpleTaskKey was executed");
```
Comment (timrijckaert) [inline example/lib/main.dart]: why?

File: example/lib/main.dart [hunk 25]
```diff
@@ -115,12 +132,16 @@ class _MyAppState extends State<MyApp> {
                       frequency: Duration(hours: 1),
                     );
                   }),
-              Text("iOS", style: Theme.of(context).textTheme.headline),
+              SizedBox(height: 16),
+              Text("Background Fetch (iOS only)", style: Theme.of(context).textTheme.headline),
```
Comment (timrijckaert) [inline example/lib/main.dart]: I would just get rid of it tbh.
I does not really belong in this library.
As an implementer I would just wanna know what is needed to use this library not how to debug it

File: README.md (hunk 20 above)
Comment (jeremie-movify) [reply README.md]: This is left to the developer's taste ; Xcode won't compile the project if either the key or its value is incorrectly formatted anyway, and this step is very straightforward for anyone experienced with Xcode

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 26]
```diff
@@ -28,27 +28,26 @@ public class SwiftWorkmanagerPlugin: FlutterPluginAppLifeCycleDelegate {
...
+        
     }
     
     public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
         
         switch call.method {
         case ForegroundMethodChannel.methods.initialize.rawValue:
             guard let callbackHandle: Int64 = call.arguments as? Int64 else {
-                result(WMPError.unexpectedMethodArguments(argumentsDescription: call.arguments.debugDescription))
+                result(WMPError.unexpectedMethodArguments(call.arguments.debugDescription))
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: Is this necessary as in, when would this crash?
This could only happen if the plugin had bad code in it, in which case the whole plugin would be rendered useless.
I would rather just forcefully cast it to an Int.

(If this is what you are doing, I'm sorry I don't read Swift so well)

File: README.md (hunk 18 above)
Comment (jeremie-movify) [reply README.md]: I agree ; I believe it should be named in another fashion.
Calling it a `Dispatcher` suggests that it actually dispatches something, which (I think) it does not ; it's definitely a Flutter entry point, as defined by :
`flutterEngine.run(withEntrypoint: flutterCallbackInformation.callbackName, libraryURI: flutterCallbackInformation.callbackLibraryPath)`

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 27]
```diff
@@ -76,60 +81,55 @@ extension SwiftWorkmanagerPlugin {
...
-                    return
+            switch call.method {
+            case BackgroundMethodChannel.methods.backgroundChannelInitialized.rawValue:
+                result(true)    // Agree to Flutter's method invocation
+                // BackgroundChannel is now available ; let's send the "iOSPerformFetch" method through it, and wait for the result
+                backgroundMethodChannel.invokeMethod(BackgroundMethodChannel.methods.iOSPerformFetch.rawValue, arguments: nil, result: { flutterResult in
+                    // We got a backgroundFetch result ; let's ensure we can convert it to a native UIBackgroundFetchResult
+                    guard
+                        let fetchResult: Int = flutterResult as? Int,
+                        let backgroundFetchResult = UIBackgroundFetchResult.init(rawValue: UInt(fetchResult))
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: 👍

File: README.md (hunk 21 above)
Comment (jeremie-movify) [reply README.md]: Yup, my bad

File: example/ios/Runner/AppDelegate.swift [hunk 28]
```diff
@@ -1,12 +1,64 @@
...
 import Flutter
+import workmanager
 
 @UIApplicationMain
 
 @objc class AppDelegate: FlutterAppDelegate {
+    
     override func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
+        
+        SwiftWorkmanagerPlugin.clearStorage()
```
Comment (timrijckaert) [inline example/ios/Runner/AppDelegate.swift]: 🤔 ? 
Explain

File: README.md (hunk 22 above)
Comment (jeremie-movify) [reply README.md]: I believe it's added value ; as a user of this plugin, I'd have a hard time figuring out how the plugin intercepts the native `performFetch`delegate call (except some Method Swizzling/dark magic thing).

File: lib/workmanager.dart [hunk 29]
```diff
@@ -100,10 +100,9 @@ class Workmanager {
       final EchoCallbackFunction echoFunction) {
     WidgetsFlutterBinding.ensureInitialized();
     _backgroundChannel.setMethodCallHandler((call) async {
-      stderr.writeln('DEBUG');
       return echoFunction(call.arguments);
```
Comment (timrijckaert) [inline lib/workmanager.dart]: make this a lambda function instead

File: example/lib/main.dart (hunk 24 above)
Comment (jeremie-movify) [reply example/lib/main.dart]: So that Flutter logs actually appear on Xcode's console when running the project from Xcode (useful for native debug purposes)

File: example/lib/main.dart (hunk 25 above)
Comment (jeremie-movify) [reply example/lib/main.dart]: I tend to disagree ; without this button (which is more a test than a debug button), there's no way for the plugin user to get his own code called, except for waiting for iOS to actually trigger `performFetchWithCompletionHandler` (which may take days).
Also, we should either provide a demo (/test) sample for both platforms or for none, not halfways :)

File: ios/Classes/SwiftWorkmanagerPlugin.swift (hunk 26 above)
Comment (jeremie-movify) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: It's more a question of best practices truly, I wouldn't force cast a value only for the sake of having to support an error scenario

File: example/ios/Runner/AppDelegate.swift (hunk 28 above)
Comment (jeremie-movify) [reply example/ios/Runner/AppDelegate.swift]: If the plugin user runs the project multiple times, nothing prevents him from tapping the "Simulate Background Fetch" button without first initialising the Flutter background service, and he wouldn't receive an error since the Callback handle has already been stored in UserDefaults through a previous call.
Clearing SwiftWorkManagerPlugin's storage ensures the app is ran from a clean state every time.

File: README.md (hunk 18 above)
Comment (timrijckaert) [reply README.md]: I took the name from the blogpost that engineered the Background entrypoint.
So while I don't fully agree that the callbackDispatcher is a good name. It is name it is referred to in Dart. 
Naming it different would make it harder to Google stuff about it.

Comment (timrijckaert) [reply README.md]: https://medium.com/flutter/executing-dart-in-the-background-with-flutter-plugins-and-geofencing-2b3e40a1a124

Search for `callbackDispatcher`

File: README.md (hunk 22 above)
Comment (timrijckaert) [reply README.md]: But you should not care as the implementer. If you care you check the code.
I try to keep the README as terse as possible

File: ios/Classes/SwiftWorkmanagerPlugin.swift (hunk 26 above)
Comment (timrijckaert) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: But it's like saying 

```
final s = "someString"

if (s is String) {
 final t = s as String;
}
```

It does not make sense. The Dart plugin will **always** send it as in `Int`

File: example/ios/Runner/AppDelegate.swift (hunk 28 above)
Comment (timrijckaert) [reply example/ios/Runner/AppDelegate.swift]: But, that's what we want right.  
It's ok for the implementer to only register the `callbackDispatcher` function once.

File: ios/Classes/SwiftWorkmanagerPlugin.swift (hunk 26 above)
Comment (jeremie-movify) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: I get your point, but if we go down this path, we could as well expect to never receive anything else than the `initialize` call methods in this channel, so we shouldn't verify the contents of `call.mehtod` either ? This doesn't look like clean code to me

File: example/ios/Runner/AppDelegate.swift (hunk 28 above)
Comment (jeremie-movify) [reply example/ios/Runner/AppDelegate.swift]: Of course, but it would cause a failure if UserDefaults is populated with a different entryPoint name (for example because the name has changed).

File: example/lib/main.dart (hunk 25 above)
Comment (timrijckaert) [reply example/lib/main.dart]: It's the same for Android.
You will need to wait.
We don't even have the ability to forcefully run the job even in debug mode.

I think the iOS code could stay in the sample but just not provided to the enduser.

The enduser can expect a fully working plugin from us

File: example/ios/Runner/AppDelegate.swift (hunk 28 above)
Comment (timrijckaert) [reply example/ios/Runner/AppDelegate.swift]: That's true! However should we catch such an error in the example?

File: ios/Classes/SwiftWorkmanagerPlugin.swift (hunk 26 above)
Comment (timrijckaert) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: Sure why not. It make total sense for the iOS code imho

Comment (jeremie-movify) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: Sorry, I cannot agree with you on this point ; not only is it dirty, but it also makes the code harder to understand. If we don't switch on the call's method to ensure we're indeed handling what we expect to handle, and we breakpoint in this method, it makes it much harder to understand where that comes from (especially since we won't have a stack trace, since this method is called by Flutter).
It also makes it harder to support adding new methods in the same channel.

Comment (timrijckaert) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: I'm not asking for you to remove the switch case. 
I just think the casting of arguments should be done safely.

File: example/ios/Runner/AppDelegate.swift (hunk 28 above)
Comment (jeremie-movify) [reply example/ios/Runner/AppDelegate.swift]: Which means you "pollute" (big quotes on this one) the SharedPrefs with something that cannot be removed ; if we push it further, it also prevents unit testing this feature correctly

Comment (timrijckaert) [reply example/ios/Runner/AppDelegate.swift]: What do you mean?

File: ios/Classes/SwiftWorkmanagerPlugin.swift (hunk 26 above)
Comment (jeremie-movify) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: ... Which it is, by "guard letting" the argument and its type ; force casting it seems like a lazy "I don't want to handle this error" move to me :-)

File: example/ios/Runner/AppDelegate.swift (hunk 28 above)
Comment (jeremie-movify) [reply example/ios/Runner/AppDelegate.swift]: Typically, you'd want to ensure that the SharedPrefs (or any other storage injected in the test) actually contains the callback handle once the foreground channel has been initialised.
The first test will pass ; but if changes are brought to the code and this callback handle is no more correctly stored in the SPs, running the test again will result in a false positive, since the SPs weren't cleaned

File: ios/Classes/SwiftWorkmanagerPlugin.swift (hunk 26 above)
Comment (timrijckaert) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: No, why would your approach be safer? 
You know because it's your own code that this can only be an `Int`.
Now you swallow the error and throw it back to the user.

If you forcefully unwrap it here and Dart did not send an Int it will crash the app, which is actually more desirable.

File: example/ios/Runner/AppDelegate.swift (hunk 28 above)
Comment (timrijckaert) [reply example/ios/Runner/AppDelegate.swift]: But that's testing? 
I don't see how this is related tbh.

You'd have some sort of cleanup of tear down of a test that's reset it to a default state.

File: ios/Classes/SwiftWorkmanagerPlugin.swift (hunk 26 above)
Comment (jeremie-movify) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: I just ran a couple scenarios ; one was force-casting the Int value sent by Dart through the MethodChannel to a String (causing a crash), the second was guard-letting it and returning a Flutter Error (as in the current implementation).
As a result, force-casting outputted "Lost connection to the device" without any clue on what caused the process to abort ; guard-letting outputted the WMPError, giving much more useful info about what caused the error.

Comment (timrijckaert) [reply ios/Classes/SwiftWorkmanagerPlugin.swift]: Only the Swift code should be changed to:

```
store(call.arguments as Int64)
```

File: README.md (hunk 18 above)
Comment (timrijckaert) [reply README.md]: Offline discussion. Name retains `callbackDispatcher` 
Some other renames will be done to make the API more clearer for starters

File: lib/workmanager.dart [hunk 30]
```diff
@@ -95,18 +95,20 @@ class Workmanager {
...
-  static void defaultCallbackDispatcher(
-      final EchoCallbackFunction echoFunction) {
+  /// A helper function so you only need to implement a [BackgroundTask]
+  static void executeTask(
+      final BackgroundTask backgroundTask) {
     WidgetsFlutterBinding.ensureInitialized();
     _backgroundChannel
-        .setMethodCallHandler((call) async => echoFunction(call.arguments));
+        .setMethodCallHandler((call) async {
+          return backgroundTask(call.arguments);
```
Comment (timrijckaert) [inline lib/workmanager.dart]: why? just use the lambda syntax

File: ios/Classes/SwiftWorkmanagerPlugin.swift [hunk 31]
```diff
@@ -1,14 +1,133 @@
...
+        backgroundMethodChannel.setMethodCallHandler { (call, result) in
+            switch call.method {
+            case BackgroundMethodChannel.methods.backgroundChannelInitialized.rawValue:
+                result(true)    // Agree to Flutter's method invocation
+                // BackgroundChannel is now available ; let's send the "iOSPerformFetch" method through it, and wait for the result
+                backgroundMethodChannel.invokeMethod(BackgroundMethodChannel.methods.iOSPerformFetch.rawValue, arguments: nil, result: { flutterResult in
+                    // We got a backgroundFetch result ; let's ensure we can convert it to a native UIBackgroundFetchResult
+                    guard
+                        let fetchResult: Int = flutterResult as? Int,
+                        let backgroundFetchResult = UIBackgroundFetchResult.init(rawValue: UInt(fetchResult))
```
Comment (timrijckaert) [inline ios/Classes/SwiftWorkmanagerPlugin.swift]: Is this Int to Enum parsing in Swift? I don't fully get what is going on here.

File: IOS_SETUP.md [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+<key>UIBackgroundModes</key>
+  <array>
+    <string>fetch</string>
+  </array>
+</key>
+```
+
+Inside your app's delegate `didFinishLaunchingWithOptions` set your desired _minimumBackgroundFetchInterval_ in your app's delegate's:
+
+//TODO: provide a complete class
```
Comment (timrijckaert) [inline IOS_SETUP.md]: Could we maybe have 1 complete example.  
Eg as an iOS noob I would not know where to place this exactly

File: lib/workmanager.dart [hunk 33]
```diff
@@ -116,22 +171,21 @@ class Workmanager {
...
+        final Duration initialDelay = _noDuration,
+        final WorkManagerConstraintConfig constraints,
+        final BackoffPolicy backoffPolicy,
+        final Duration backoffPolicyDelay = _noDuration,
+      }) async =>
       await _register(
         methodName: "registerOneOffTask",
         uniqueName: uniqueName,
-        echoValue: echoValue,
+        echoValue: valueToReturn,
```
Comment (timrijckaert) [inline lib/workmanager.dart]: Ideally this should also be renamed on native, but let's leave it for now.
